app/src/main/java/com/gruppo3/wetravel/mapManager/DirectionsManager.java

```diff
@@ -0,0 +1,122 @@
+package com.gruppo3.wetravel.mapManager;
+
+import android.graphics.Color;
+
+import com.google.android.gms.maps.GoogleMap;
+import com.google.android.gms.maps.model.LatLng;
+import com.google.android.gms.maps.model.PolylineOptions;
+import com.gruppo3.wetravel.mapManager.asyncTasks.DownloadTask;
+import com.gruppo3.wetravel.mapManager.asyncTasks.ParserTask;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+/**
+ * Singleton class that downloads JSON data from Google Maps Directions Web Service and shows in a map fragment
+ *
+ * @author Giovanni Barca
+ */
+public class DirectionsManager {
+
+    public enum DirectionModes {
+        DRIVING,
+        WALKING,
+        BICYCLING,
+        TRANSIT
+    }
+
+    private static final DirectionsManager instance = new DirectionsManager();
+
+    /**
+     * Gets the instance of this singleton class.
+     * @return Instance of this singleton class
+     */
+    public static DirectionsManager getInstance() {
+        return instance;
+    }
+
+    /**
+     * Computes and shows on the map the requested route from origin to destination.
+     * @param origin Route's origin coordinates
+     * @param dest Route's destination coordinates
+     */
+    public void computeRoute(final GoogleMap googleMap, LatLng origin, LatLng dest, DirectionModes directionMode) {
+        String googleDirectionsApiUrl = buildDirectionsUrl(origin, dest, directionMode); // Builds the Google Maps Directions API request url
+
+        // Downloads the JSON from Google Maps Directions Web Service, parses it and sends to showRoutFromPrasedJSON(String) if there were no errors
+        new DownloadTask(new DownloadTask.AsyncResponse() {
+            @Override
+            public void onFinishResult(String result) {
+                // Continue if download successfully terminated
+                if (!result.equals(DownloadTask.ERROR_MESSAGE)) {
+                    new ParserTask(new ParserTask.AsyncResponse() {
+                        @Override
+                        public void onFinishResult(List<List<HashMap<String, String>>> result) {
+                            if (result.size() > 0) // If there were errors, list size is 0
```

Comment:
You might consider `isEmpty()` method...
```suggestion
                            if (!result.isEmpty()) // If there were errors, list is empty.
```